File: packages/button-ios/Sources/Button/PSButton.swift
import SwiftUI

public struct PSButton: View {
    public let label: String
    public let variant: PSButtonVariant
    public let size: PSButtonSize
    public let isDisabled: Bool
    public let isLoading: Bool
    public let fullWidth: Bool
    public let iconLeading: Image?
    public let iconTrailing: Image?
    public let accessibilityLabel: String?
    public let action: () -> Void

    @StateObject private var viewModel: PSButtonViewModel

    public init(
        label: String,
        variant: PSButtonVariant = .primary,
        size: PSButtonSize = .md,
        isDisabled: Bool = false,
        isLoading: Bool = false,
        fullWidth: Bool = false,
        iconLeading: Image? = nil,
        iconTrailing: Image? = nil,
        accessibilityLabel: String? = nil,
        action: @escaping () -> Void
    ) {
        precondition(!label.trimmingCharacters(in: .whitespaces).isEmpty,
                     "PSButton: label must not be empty")
        self.label = label
        self.variant = variant
        self.size = size
        self.isDisabled = isDisabled
        self.isLoading = isLoading
        self.fullWidth = fullWidth
        self.iconLeading = iconLeading
        self.iconTrailing = iconTrailing
        self.accessibilityLabel = accessibilityLabel
        self.action = action
        _viewModel = StateObject(wrappedValue: PSButtonViewModel(variant: variant, size: size))
    }

    private var tokens: PSButtonTokenSet {
        PSButtonViewModel.resolveTokenSet(
            variant: variant,
            state: (isDisabled || isLoading) ? .disabled : .default
        )
    }

    public var body: some View {
        Button(action: {
            viewModel.handlePress(isDisabled: isDisabled, isLoading: isLoading, action: action)
        }) {
            HStack(spacing: size.gap) {
                if isLoading {
                    ProgressView()
                        .progressViewStyle(.circular)
                        .tint(Color(tokens.foregroundColor))
                        .frame(width: size.iconSize, height: size.iconSize)
                } else if let icon = iconLeading {
                    icon.resizable()
                        .frame(width: size.iconSize, height: size.iconSize)
                }

                Text(label)
                    .font(.system(size: size.fontSize, weight: PSButtonTokens.fontWeight))

                if !isLoading, let icon = iconTrailing {
                    icon.resizable()
                        .frame(width: size.iconSize, height: size.iconSize)
                }
            }
            .foregroundColor(Color(tokens.foregroundColor))
            .padding(.horizontal, size.paddingX)
            .padding(.vertical, size.paddingY)
            .frame(maxWidth: fullWidth ? .infinity : nil)
            .frame(height: size.height)
            .background(Color(tokens.backgroundColor))
            .overlay(
                RoundedRectangle(cornerRadius: PSButtonTokens.radius)
                    .stroke(Color(tokens.borderColor), lineWidth: tokens.borderWidth)
            )
            .cornerRadius(PSButtonTokens.radius)
        }
        .disabled(isDisabled || isLoading)
        .accessibilityLabel(accessibilityLabel ?? label)
        .accessibilityAddTraits(.isButton)
        .accessibilityValue(isLoading ? "Loading" : "")
    }
}

File: packages/button-android/src/main/kotlin/com/porto/ds/button/PSButton.kt
package com.porto.ds.button

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.disabled
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.stateDescription
import androidx.compose.ui.unit.dp
import androidx.lifecycle.viewmodel.compose.viewModel

@Composable
fun PSButton(
    label: String,
    onPress: () -> Unit,
    modifier: Modifier = Modifier,
    variant: ButtonVariant = ButtonVariant.Primary,
    size: ButtonSize = ButtonSize.MD,
    isDisabled: Boolean = false,
    isLoading: Boolean = false,
    fullWidth: Boolean = false,
    iconLeading: ImageVector? = null,
    iconTrailing: ImageVector? = null,
    accessibilityLabel: String? = null,
    vm: PSButtonViewModel = viewModel(),
) {
    require(label.isNotBlank()) { "PSButton: label must not be blank" }

    val effectivelyDisabled = isDisabled || isLoading
    val uiState by vm.uiState.collectAsState()

    LaunchedEffect(variant, size, isDisabled, isLoading) {
        vm.processIntent(ButtonIntent.Configure(variant, size, isDisabled, isLoading))
    }

    val t = uiState.tokenSet
    val sz = size.tokens
    val shape = RoundedCornerShape(PSButtonTokens.Radius)

    val border =
        if (variant is ButtonVariant.Secondary) {
            BorderStroke(PSButtonTokens.BorderWidth, t.borderColor)
        } else {
            null
        }

    val rowModifier =
        modifier
            .defaultMinSize(minWidth = 44.dp, minHeight = 44.dp)
            .height(sz.height)
            .then(if (fullWidth) Modifier.fillMaxWidth() else Modifier)
            .semantics {
                role = Role.Button
                contentDescription = accessibilityLabel ?: label
                if (effectivelyDisabled) disabled()
                if (isLoading) stateDescription = "Loading"
            }

    Button(
        onClick = { vm.processIntent(ButtonIntent.Press, onPress = onPress) },
        modifier = rowModifier,
        enabled = !effectivelyDisabled,
        shape = shape,
        colors =
            ButtonDefaults.buttonColors(
                containerColor = t.backgroundColor,
                contentColor = t.foregroundColor,
                disabledContainerColor = t.backgroundColor,
                disabledContentColor = t.foregroundColor,
            ),
        border = border,
        contentPadding = PaddingValues(horizontal = sz.paddingX, vertical = sz.paddingY),
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(sz.gap),
        ) {
            if (isLoading) {
                CircularProgressIndicator(
                    modifier = Modifier.size(sz.iconSize),
                    color = t.foregroundColor,
                    strokeWidth = 2.dp,
                )
            } else if (iconLeading != null) {
                Icon(
                    imageVector = iconLeading,
                    contentDescription = null,
                    modifier = Modifier.size(sz.iconSize),
                )
            }

            Text(
                text = label,
                fontSize = sz.fontSize,
                fontWeight = PSButtonTokens.fontWeight,
            )

            if (!isLoading && iconTrailing != null) {
                Icon(
                    imageVector = iconTrailing,
                    contentDescription = null,
                    modifier = Modifier.size(sz.iconSize),
                )
            }
        }
    }
}
